src/vs/workbench/contrib/positronHoloViews/browser/positronHoloViewsService.ts

```diff
@@ -0,0 +1,169 @@
+ *  Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+ *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
+ *--------------------------------------------------------------------------------------------*/
+import { Disposable, DisposableStore } from 'vs/base/common/lifecycle';
+import { Emitter, Event } from 'vs/base/common/event';
+import { ILanguageRuntimeMessageOutput, LanguageRuntimeSessionMode, RuntimeOutputKind } from 'vs/workbench/services/languageRuntime/common/languageRuntimeService';
+import { NotebookOutputPlotClient } from 'vs/workbench/contrib/positronPlots/browser/notebookOutputPlotClient';
+import { IPositronHoloViewsService, MIME_TYPE_HOLOVIEWS_EXEC } from 'vs/workbench/services/positronHoloViews/common/positronHoloViewsService';
+import { ILanguageRuntimeSession, IRuntimeSessionService } from 'vs/workbench/services/runtimeSession/common/runtimeSessionService';
+import { InstantiationType, registerSingleton } from 'vs/platform/instantiation/common/extensions';
+import { IPositronNotebookOutputWebviewService } from 'vs/workbench/contrib/positronOutputWebview/browser/notebookOutputWebviewService';
+
+const MIME_TYPE_HTML = 'text/html';
+const MIME_TYPE_PLAIN = 'text/plain';
+
+export class PositronHoloViewsService extends Disposable implements IPositronHoloViewsService {
+	/** Needed for service branding in dependency injector. */
+	_serviceBrand: undefined;
+
+	/** Placeholder for service initialization. */
+	initialize() { }
+
+	/** Map of holoviz messages keyed by session ID. */
+	private readonly _messagesBySessionId = new Map<string, ILanguageRuntimeMessageOutput[]>();
+
+	/**
+	 * Map to disposeable stores for each session. Used to preventing memory leaks caused by
+	 * repeatedly attaching to the same session which can happen in the case of the application
+	 * closing before the session ends
+	 */
+	private _sessionToDisposablesMap = new Map<string, DisposableStore>();
+
+	/** The emitter for the onDidCreatePlot event */
+	private readonly _onDidCreatePlot = this._register(new Emitter<NotebookOutputPlotClient>());
+
+	/** Emitted when a new HoloViews webview is created. */
+	onDidCreatePlot: Event<NotebookOutputPlotClient> = this._onDidCreatePlot.event;
+
+	constructor(
+		@IRuntimeSessionService private _runtimeSessionService: IRuntimeSessionService,
+		@IPositronNotebookOutputWebviewService private _notebookOutputWebviewService: IPositronNotebookOutputWebviewService,
+	) {
+		super();
+
+		// Attach to existing sessions.
+		this._runtimeSessionService.activeSessions.forEach(session => {
+			this._attachSession(session);
+		});
+
+		// Attach to new sessions.
+		this._register(this._runtimeSessionService.onWillStartSession((event) => {
+			this._attachSession(event.session);
+		}));
+	}
+
+	override dispose(): void {
+		super.dispose();
+		// Clean up disposables linked to any connected sessions
+		this._sessionToDisposablesMap.forEach(disposables => disposables.dispose());
```

Comment:
Not sure if its necessary, but should we clear the map too?

Reply:
Probably!

Reply:
Actually, I guess this will automatically get garbage collected. Assuming `dispose()` only gets called when the service is truly being disposed of.